DRILL-7038: Queries on partitioned columns scan the entire datasets

KazydubB opened the pull request:
Added new optimizer rule which checks if query references directory columns only and has DISTINCT or GROUP BY operation. If the condition holds, instead of scanning full file set the following will be performed:
- if there is cache metadata file, these directories will be read from it,
- otherwise directories will be gathered from selection object (PartitionLocation).

In the end Scan node will be transformed to DrillValuesRel (containing constant literals) with gathered values so no "scan" will be performed.

priteshm commented:
@gparai @HanumathRao can you'll review?

gparai commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java:
```diff
@@ -550,4 +567,180 @@ private static void setPruneStatus(MetadataContext metaContext, PruneStatus prun
+          if (index < parts.size()) {
+            values.add(parts.get(index));
+          } else {
+            values.add(null);
```
Why do we add null here?

KazydubB replied:
Added comment.

gparai commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java:
```diff
@@ -550,4 +567,180 @@ private static void setPruneStatus(MetadataContext metaContext, PruneStatus prun
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
```
Please add a comment describing what this transformation does maybe with an example.

KazydubB commented:
@gparai thank you, added comments.

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java:
```diff
@@ -550,4 +567,211 @@ private static void setPruneStatus(MetadataContext metaContext, PruneStatus prun
+
+    private List<String> getValues(String selectionRoot, List<String> directories, List<Integer> indexes) {
+      List<String> values = new ArrayList<>();
+      for (String dir : directories) {
```
Consider 
```suggestion
    directories.stream()
        .map(dir -> ColumnExplorer.listPartitionValuesDirectory(dir, selectionRoot))
        .forEachOrdered(parts -> indexes.stream()
            .map(index -> index < parts.size() ? parts.get(index) : null)
            .forEach(values::add));
```
```

KazydubB replied:
Thanks for suggestion but I find it harder to read.

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/ColumnExplorer.java:
```diff
@@ -224,6 +224,33 @@ public static boolean isPartitionColumn(String partitionDesignator, String path)
     return Arrays.asList(diffDirectoryNames);
   }
 
+  public static List<String> listPartitionValuesDirectory(String dirPath, String root) {
```
It copies 90% of `listPartitionValues()` method. Could you just pass some boolean flag to the old method, for example   `hasDirsOnly` as [here](https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/planner/DFSFilePartitionLocation.java#L29)?

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java:
```diff
@@ -550,4 +567,211 @@ private static void setPruneStatus(MetadataContext metaContext, PruneStatus prun
+
+    private List<String> getValues(List<PartitionLocation> partitions, List<Integer> indexes) {
+      List<String> values = new ArrayList<>(partitions.size() * indexes.size());
+      for (PartitionLocation partition : partitions) {
```
Why do not use Stream API?

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/TestPartitionPruning.java:
```diff
@@ -0,0 +1,73 @@
+
+  @Test
+  public void testPartitionPruning() throws Exception {
+    String[] tableNames = {"multilevel/parquet", "multilevel/json", "multilevel/csv"};
```
Interesting that we had not Partition Pruning for the CSV or JSON formats earlier.
Could elaborate how to update Drill doc ["Partition Pruning Introduction"](https://drill.apache.org/docs/partition-pruning-introduction/) in accordance to your improvements?

KazydubB commented:
@vdiravka currently we can select partition columns for tables consisting of many directories containing JSON/csv as well. There is no need to update documentation as this feature does not add new functionality but only optimizes existing one. I see that the confusion may arise from bad naming - hence I renamed the test class with its methods and also renamed the rule class.

Resolved the conflicts. Please, take a look once more.

gparai commented:
LGTM +1

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/TestPartitionColumnScanPruning.java:
```diff
@@ -0,0 +1,73 @@
+import java.nio.file.Paths;
+
+@Category({PlannerTest.class})
+public class TestPartitionColumnScanPruning extends PlanTestBase {
```
Please add also the test case for Hive table with partitions, for instance one test case in the `TestHiveStorage`

KazydubB replied:
This change targets file system databases only. Hive tables do not support querying of partition columns (dir0, dir1 etc.).

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/TestPartitionColumnScanPruning.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  @Test
+  public void testPartitionColumnScanPruning() throws Exception {
```
```suggestion
  public void testDirScanToValuesConverting() throws Exception {
```

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/TestPartitionColumnScanPruning.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  @Test
+  public void testPartitionColumnScanPruningWithMetadataCache() throws Exception {
```
```suggestion
  public void testDirScanToValuesConvertingWithMetadataCache() throws Exception {
```

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/TestPartitionColumnScanPruning.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  @Test
+  public void testPartitionColumnScanPruningIsNotApplied() throws Exception {
```
```suggestion
  public void testDirScanToValuesConvertingIsNotApplied() throws Exception {
```

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/TestPartitionColumnScanPruning.java:
```diff
@@ -0,0 +1,73 @@
+import java.nio.file.Paths;
+
+@Category({PlannerTest.class})
+public class TestPartitionColumnScanPruning extends PlanTestBase {
```
```suggestion
public class TestDirScanToValuesConverting extends PlanTestBase {
```

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/TestDirScanToValuesConverting.java:
```diff
@@ -0,0 +1,73 @@
+import java.nio.file.Paths;
+
+@Category({PlannerTest.class})
+public class TestDirScanToValuesConverting extends PlanTestBase {
```
Please make class name to be consistent with test names.
```suggestion
public class TestDirScanToValuesConversion extends PlanTestBase {
```
